packages/hdwallet-provider/src/index.ts

```diff
@@ -37,50 +158,52 @@ class HDWalletProvider {
   public engine: ProviderEngine;
 
   constructor(
-    mnemonic: string | string[],
-    provider: string | any,
-    addressIndex: number = 0,
-    numAddresses: number = 10,
-    shareNonce: boolean = true,
-    walletHdpath: string = `m/44'/60'/0'/0/`
+    mnemonicOrPrivateKeysOrOptions: string | string[] | ConstructorOptions,
+    ...args: any[]
   ) {
-    this.walletHdpath = walletHdpath;
+    const {
+      mnemonic,
+      privateKeys,
+      url,
+      addressIndex,
+      numberOfAddresses,
+      sharedNonce,
+      derivationPath
+    } = normalizeConstructorOptions(mnemonicOrPrivateKeysOrOptions, args);
+
+    this.walletHdpath = derivationPath;
     this.wallets = {};
     this.addresses = [];
     this.engine = new ProviderEngine();
 
-    if (!HDWalletProvider.isValidProvider(provider)) {
+    if (!HDWalletProvider.isValidProvider(url)) {
```

Comment:
I just realized that `isValidProvider` is a pretty awful method. We should probably update that at some point. Or else we should just rename it something like `hasValidProtocol`.

Reply:
I'm open to renaming if you want me too (while we're here cleaning things up anyway).

Reply:
Note: this just turned into `providerOrUrl`, because that's what the types suggested previously (`string | any`!)